Kernel/Storage/ATA/AHCIController.cpp

```diff
@@ -26,6 +26,28 @@ UNMAP_AFTER_INIT NonnullRefPtr<AHCIController> AHCIController::initialize(PCI::D
 
 bool AHCIController::reset()
 {
+    SpinlockLocker locker(m_control);
+    Array<u32, 32> lock_flags {};
+
+    // Note: Since we do a full AHCI reset here, we need to prevent individual AHCI ports
+    // from trying to initiate a transaction as this can be fatal when doing full AHCI
+    // controller reset. We keep the transaction initiation logic in each AHCIPort
+    // and not putting it in the AHCIController as it will impact performance of
+    // other attached devices severely, but we can still use the hard Spinlock of each port
+    // to ensure nothing really bad happens during this reset sequence.
+
+    for (auto port : m_ports) {
+        if (port)
+            lock_flags[port->m_port_index] = port->m_hard_lock.lock();
+    }
+
+    ScopeGuard release_all_spinlocks([this, lock_flags]() {
+        for (auto port : m_ports) {
+            if (port)
+                port->m_hard_lock.unlock(lock_flags[port->m_port_index]);
+        }
+    });
```

Comment:
This seems really flimsy and I'm referring both to this code, and the rest of the AHCIPort locking code. It looks like the AHCIPort has 2 locks, a spinlock and a mutex, but it locks both of them every time, aside from a singular instance (inside a workqueue item) where we only lock the mutex, what's the point of that? If the point of using a mutex is to allow threads to yield while waiting, why use a spinlock? If the point of using the spinlock is to not allow blocking in interrupts, why use a mutex everywhere?
And about this code specifically, this seems like a really invasive/ugly way to this and it also seems a bit flimsy, assuming we did lock their spinlocks, and now they're spinning waiting until we're done, is it even valid for them to continue their operation when we just reset the controller?

Reply:
The idea behind using both a spinlock and a mutex is to allow the Work queue item to use the AHCI port, so after a DMA transaction, it will lock the mutex so no other thread can use it in that period of time, but we don't take the spinlock because we don't do anything really critical (i.e. changing hardware state). When we need to use the AHCI port to do DMA transfer, we lock both to ensure that we don't allow other threads to use the AHCI port but also to ensure we won't get out of scope until we are done changing the hardware state to initiate a DMA transfer (and also that no other CPU can use the AHCI port other than the current CPU). We take the spinlock in the AHCI controller reset code because it's actually OK to allow a Work Queue item to handle the result of a DMA transfer (so it can take the mutex and move the data from the DMA buffer), but actually trying to change the hardware state is a bad thing according to the spec while resetting the entire HBA.

Reply:
I could add this to the document I wrote here:
https://github.com/SerenityOS/serenity/blob/master/Documentation/Kernel/AHCILocking.md
if needed :)